### PR TITLE
fix(fibre)!: signature verification bug

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Overview

Fixes https://linear.app/celestia/issue/PROTOCO-2112/unsettled-msgpayforfibre-commits-its-fibre-system-blob-into-the-data

### CODE REVIEW

File: app/pay_for_fibre_validation.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+// extractPayForFibre returns the tx's MsgPayForFibre, or nil if the tx
+// contains none. It errors if a PFF is mixed with other messages or
+// duplicated — the same rule ProcessProposal enforces — since such a tx can
+// never be included in a valid block.
+func extractPayForFibre(tx sdk.Tx) (*fibretypes.MsgPayForFibre, error) {
+	msgs := tx.GetMsgs()
+	if len(msgs) == 0 {
+		return nil, nil
```
Comment (rach-id) [inline app/pay_for_fibre_validation.go]: the `nil,nil` return most of the times causes nil panics after some time because when `err ==nil`, mostly you use the value directly if you don't have this context. So, if it's possible to return an error here, would be better

File: app/pay_for_fibre_validation.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	var pff *fibretypes.MsgPayForFibre
+	for _, msg := range msgs {
+		if m, ok := msg.(*fibretypes.MsgPayForFibre); ok {
+			pff = m
+		}
```
Comment (rach-id) [inline app/pay_for_fibre_validation.go]: probably we could break early after finding the PFF because we know the transcation contains a PFF. So the only thing left check is `len(msgs) != 1`

File: app/pay_for_fibre_validation.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	if pff != nil && len(msgs) != 1 {
+		return nil, sdkerrors.Wrapf(apperr.ErrInvalidPayForFibreTx, "tx contains a MsgPayForFibre and %d total messages", len(msgs))
+	}
+	return pff, nil
+}
+
+func pffSignatureVerificationCacheKey(tx []byte) [sha256.Size]byte {
```
Comment (rach-id) [inline app/pay_for_fibre_validation.go]: This means we will do a hash operation on every call to get the key. If this proves to be expensive CPU wise, we could explore how to optimise afterwards

File: app/pay_for_fibre_validation.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	return sha256.Sum256(tx)
+}
+
+func (app *App) IsPayForFibreSignatureVerificationCached(tx []byte) bool {
+	_, ok := app.pffSignatureVerificationCache.Load(pffSignatureVerificationCacheKey(tx))
+	return ok
+}
+
+func (app *App) CachePayForFibreSignatureVerification(tx []byte) {
+	app.pffSignatureVerificationCache.Store(pffSignatureVerificationCacheKey(tx), struct{}{})
```
Comment (rach-id) [inline app/pay_for_fibre_validation.go]: [separate issue]
what about transactions that get evicted from the mempool due to TTL? to avoid having an infinitely growing cache, we could adopt some eviction policy that matches the mempool's one so we evict transactions that never get included.
This might apply to PFBs cache as well

File: x/fibre/ante/ante.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+) FibreSignatureDecorator {
+	return FibreSignatureDecorator{
+		keeper:               keeper,
+		isVerificationCached: isVerificationCached,
+		cacheVerification:    cacheVerification,
+	}
+}
+
+func (d FibreSignatureDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if simulate {
```
Comment (rach-id) [inline x/fibre/ante/ante.go]: Hum, this is tricky. Simulation should also return the gas used in ante handler. And given most of the cost is from signature verification, if we skip here, we wouldn't be reporting that gas => invalid gas simulation.

I'm also thinking how we should handle the case where we have a cache hit and how to account for that gas, because it's work that the CPU did during checkTx.

I'm not sure how we should do this, but this is an important thing that needs to be addressed

**Review by rach-id — COMMENTED** [review]
The PR is directionaly right. Left some comments that probably you already thought of, but left them to remind myself to check them in the future

File: app/pay_for_fibre_validation.go (hunk 4 above)
Comment (ninabarbakadze) [reply app/pay_for_fibre_validation.go]: Yes for sure! Also noticed there's no pruning in PFB cache which is what this one is mirroring. Let's add a pruning interval or something easy in a FLUP to both of them.

File: x/fibre/ante/ante.go (hunk 5 above)
Comment (ninabarbakadze) [reply x/fibre/ante/ante.go]: I think we should charge the same sig verification gas on cache hits, misses, and simulation. We don’t currently have a fibre specific formula so we need to define one. For example the number of validator signatures multiplied by the SDK’s SigVerifyCostED25519. Charge it upfront then skip the actual crypto on a cache hit. This should keep gas deterministic and save us some CPU.

**Comment by ninabarbakadze** [comment]
how do i disable copilot review on every push it's very noisy 😭

**Comment by ninabarbakadze** [comment]
I think we can do the second round of review cc @rach-id I resolved all of your comments just lost track of in which commit because I made many since then so not linking specific commits in responses if that's ok 🙏🏻

I will open follow-up issue for benchmarking required gas and tuning the numbers.

I'd love to get feedback on structure + test coverage.

Related bug to entries to being pruned was already reported https://linear.app/celestia/issue/PROTOCO-2076/blobtx-validation-cache-can-grow-unbounded-medium we can groom it to include the pff cache

**Codex flagged this:** 

`ProcessProposal` performs stateful PFF validation but does not apply the resulting Fibre state changes. Because successful checks neither deduct escrow nor mark promises as processed, subsequent PFFs are validated against unchanged Fibre state. Distinct transactions containing the same promise, or promises whose combined cost exceeds escrow, can therefore pass `ProcessProposal` but fail during `FinalizeBlock` after their Fibre system blobs are already committed. Apply each successful PFF transition to the cached proposal context before validating the next transaction, and add tests for duplicate promises and cumulative escrow exhaustion.

**Regarding test coverage suggestions were:**

The critical missing cases are multiple PFFs in one proposal: duplicate promises, cumulative escrow exhaustion, PrepareProposal filtering of statefully invalid PFFs, and cache entries that never finalize. Overall, I’d consider the feature partially covered—not sufficient for the main regression risk yet.

**Review by ninabarbakadze — COMMENTED** [review]
Here's claude review findings so we don't have to waste credits re-running it

Most severe:
1. Escrow overdraw / duplicate promises still commit blobs (process_proposal.go:121) — ProcessProposal validates every promise against static pre-block state, so N txs against one escrow all pass while collectively overdrawing; blob data is already in the square. This is the exact bypass the PR meant to close.

2. Router paths skip verification entirely (msg_server.go:124) — a gov proposal carrying MsgPayForFibre settles a promise with zero signature checks.

@ninabarbakadze  This is related to us not running validator sig verification in FinalieBlock. IIRC we made this decision specifically but obviously can add a cache check. cc @rach-id 

3. ~~Multi-message txs skip both ante decorators (ante.go:74) — no verification anywhere at execution for such txs.~~

@ninabarbakadze - those decorators are for PFFs which should not have more than one message anyways

4. Liveness stall (process_proposal.go:121) — stale PFF txs never evicted by recheck get proposed, then every validator rejects the whole block.

5. ~~Unbounded cache memory leak / DoS (pff_tx_cache.go:33) — never-included txs leak sync.Map entries forever.~~

@ninabarbakadze we are aware of this

File: pkg/appconsts/fibre_gas_consts.go [hunk 6]
```diff
@@ -13,4 +13,12 @@ const (
 
 	// PFBFibreChunkSize is the chunk size (256 KiB) used for gas calculation in PayForFibre.
 	PFBFibreChunkSize uint32 = 262_144
+
+	// PFBFibreTxGasFixedCost is the fixed gas for MsgPayForFibre validation.
+	// TODO: benchmark and tune.
```
Comment (rach-id) [inline pkg/appconsts/fibre_gas_consts.go:18]: Do we have issues for these?

File: x/fibre/keeper/keeper.go [hunk 7]
```diff
@@ -16,10 +16,11 @@ import (
...
 type Keeper struct {
-	cdc           codec.Codec
-	storeKey      storetypes.StoreKey
-	bankKeeper    types.BankKeeper
-	stakingKeeper types.StakingKeeper
+	cdc                     codec.Codec
+	storeKey                storetypes.StoreKey
+	bankKeeper              types.BankKeeper
+	stakingKeeper           types.StakingKeeper
+	pffSigVerificationCache *sigVerificationCache
```
Comment (rach-id) [inline x/fibre/keeper/keeper.go]: I don't think this should live here. The cache is not committed to the state. IMO, we should have the cache in the `app` package next to where it's used: abci methods

File: x/fibre/types/gas.go [hunk 8]
```diff
@@ -25,6 +25,12 @@ func EstimateGasForPayForFibre(blobSize uint32) uint64 {
 	return appconsts.PFBFibreGasFixedCost + appconsts.PFBFibreGasPerChunk*chunks
 }
 
+// EstimateGasForPayForFibreSignatureVerification returns fixed plus per signature ante gas.
+func EstimateGasForPayForFibreSignatureVerification(validatorSignatureCount uint64) uint64 {
+	return appconsts.PFBFibreTxGasFixedCost +
```
Comment (rach-id) [inline x/fibre/types/gas.go]: the `B` is extra here: PF**B**FibreTxGasFixedCost. same for `PFBFibreSignatureGasPerValidatorSignature`

File: app/upgrades.go [hunk 9]
```diff
@@ -30,7 +30,7 @@ import (
 )
 
 // RegisterUpgradeHandlers is used for registering any on-chain upgrades.
-func (app App) RegisterUpgradeHandlers() {
+func (app *App) RegisterUpgradeHandlers() {
```
Comment (rach-id) [inline app/upgrades.go]: why?

File: x/fibre/ante/ante.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+func (FibreSignatureGasDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if msg := payForFibreMessage(tx); msg != nil {
+		consumeDeterministicFibreSignatureGas(ctx, msg)
+	}
+	return next(ctx, tx, simulate)
+}
+
+func consumeDeterministicFibreSignatureGas(ctx sdk.Context, msg *fibretypes.MsgPayForFibre) {
+	ctx.GasMeter().ConsumeGas(
+		fibretypes.EstimateGasForPayForFibreSignatureVerification(uint64(len(msg.ValidatorSignatures))),
```
Comment (rach-id) [inline x/fibre/ante/ante.go:30]: We don't verify all the signatures, only the first 2/3rds only if we don't encounter any invalid signature. So, if we keep it like this, we're overpaying. IMO, we should only provide `2/3rd+1` signature count

**Review by rach-id — COMMENTED** [review]
LGTM 👍  I didn't review the tests, only the implementation. Once that part is good, I'll review the tests.

> Escrow overdraw / duplicate promises still commit blobs (process_proposal.go:121) — ProcessProposal validates every promise against static pre-block state, so N txs against one escrow all pass while collectively overdrawing; blob data is already in the square. This is the exact bypass the PR meant to close.

This is the interesting one, IMO, we should execute these transactions on the prepare/process temporary state. But should be FLUP to be reviewed separately. Otherwise, we stack it on top of this PR, using github PR stacks

> Router paths skip verification entirely (msg_server.go:124) — a gov proposal carrying MsgPayForFibre settles a promise with zero signature checks.

This is fine, we can do a FLUP where we don't allow gov to send PFFs or PFBs, but that's very unlikely to happen anyway.

> Liveness stall (process_proposal.go:121) — stale PFF txs never evicted by recheck get proposed, then every validator rejects the whole block.

IMO, this should be a FLUP and we do the staleness check in prepare/process proposal. For CheckTx, I think it's fine, these transactions can be cleaned using TTLs.

File: x/fibre/keeper/keeper.go (hunk 7 above)
Comment (ninabarbakadze) [reply x/fibre/keeper/keeper.go]: It was in the app and then I moved it because it was mostly used by the keeper and i was routing the functions through ante.

Comment (rach-id) [reply x/fibre/keeper/keeper.go]: got you 👍 I think it should be moved back there where it belongs. Otherwise, we're adding more things to the keeper that are not committed and can lead to confusion in the future

File: pkg/appconsts/fibre_gas_consts.go (hunk 6 above)
Comment (ninabarbakadze) [reply pkg/appconsts/fibre_gas_consts.go:18]: https://github.com/celestiaorg/celestia-app/issues/7628

File: x/fibre/keeper/keeper.go (hunk 7 above)
Comment (ninabarbakadze) [reply x/fibre/keeper/keeper.go]: done! https://github.com/celestiaorg/celestia-app/pull/7599/changes/4c557f64000262c5768b7d13bdd0619ef61490d2

File: x/fibre/types/gas.go (hunk 8 above)
Comment (ninabarbakadze) [reply x/fibre/types/gas.go]: renamed and simplified https://github.com/celestiaorg/celestia-app/pull/7599/changes/1f5afde759528d8da9a8880dec85e14682e8671f

File: app/upgrades.go (hunk 9 above)
Comment (ninabarbakadze) [reply app/upgrades.go]: go vet gave me failures before because i changed something in app but i think it's no longer the case. Reverted https://github.com/celestiaorg/celestia-app/pull/7599/commits/21b350ed288140a3aca5eb136f2d43c9f2e289de

File: x/fibre/ante/ante.go (hunk 10 above)
Comment (ninabarbakadze) [reply x/fibre/ante/ante.go:30]: should we just hardcode 2/3 of the 100 validators celestia allows + 1?

**Comment by ninabarbakadze** [comment]
On main all the PFF sateful + val signature checks lived in the `PayForFibre` msg server, which only runs at `FinalizeBlock`. This PR spread them across the ante handler + proposal phases and dropped validator sig verification from `PayForFibre`.

```
┌──────────────────────────────────────────────────┬───────────────────┬────────────────────────────────────┬─────────────────────────────────────┐
│                      Check                       │      Before       │               After                │       Still gates settlement?       │
├──────────────────────────────────────────────────┼───────────────────┼────────────────────────────────────┼─────────────────────────────────────┤
│ Stateless pp.Validate() (signer sig, formats)    │ msg server        │ ante + msg server                  │ ✅ still in msg server              │
├──────────────────────────────────────────────────┼───────────────────┼────────────────────────────────────┼─────────────────────────────────────┤
│ Validator sig verify (2/3)                       │ msg server        │ ante only                          │ ⚠️ only if ante runs + cache misses │
├──────────────────────────────────────────────────┼───────────────────┼────────────────────────────────────┼─────────────────────────────────────┤
│ Stateful check (balance, not-processed, windows) │ msg server        │ ProcessProposal (new) + msg server │ ✅ still in msg server              │
├──────────────────────────────────────────────────┼───────────────────┼────────────────────────────────────┼─────────────────────────────────────┤
│ Tx shape (1 msg only)                            │ Process + Prepare │ CheckTx + Process + Prepare        │ ❌ not at FinalizeBlock             │
├──────────────────────────────────────────────────┼───────────────────┼────────────────────────────────────┼─────────────────────────────────────┤
│ Max PFF / block                                  │ Process + Prepare │ Process + Prepare                  │ ❌ consensus-only (fine)            │
└──────────────────────────────────────────────────┴───────────────────┴────────────────────────────────────┴─────────────────────────────────────┘
```

TL;DR: stateless + stateful checks are still in the msg server, so no regression there. The one that is no longer run is the validator sig verification.

File: pkg/appconsts/fibre_gas_consts.go (hunk 6 above)
Comment (ninabarbakadze) [reply pkg/appconsts/fibre_gas_consts.go:18]: also:

https://github.com/celestiaorg/celestia-app/issues/7629 
https://github.com/celestiaorg/celestia-app/issues/7630
https://github.com/celestiaorg/celestia-app/issues/7631

File: x/fibre/ante/ante.go (hunk 10 above)
Comment (rach-id) [reply x/fibre/ante/ante.go:30]: 100 is not hardcoded, it can be changed by gov. So probably would be a good idea to get it from there

**Review by rach-id — APPROVED** [review]
Great stuff 🎉
